vcd/resource_generic_crud.go

```diff
@@ -0,0 +1,192 @@
+	preDeleteHooks []outerEntityHook[O]
+
+	// readHooks that will be executed after the entity is read, but before it is stored in state
+	readHooks []outerEntityHook[O]
+}
+
+// updateDeleter is a type constraint to match only entities that have Update and Delete methods
+type updateDeleter[O any, I any] interface {
+	Update(*I) (O, error)
+	Delete() error
+}
+
+// outerEntityHook defines a type for hook that can be fed into generic CRUD operations
+type outerEntityHook[O any] func(O) error
+
+// schemaHook defines a type for hook that can be fed into generic CRUD operations
+type schemaHook func(*VCDClient, *schema.ResourceData) error
+
+func createResource[O updateDeleter[O, I], I any](ctx context.Context, d *schema.ResourceData, meta interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	t, err := c.getTypeFunc(d)
+	if err != nil {
+		return diag.Errorf("error getting %s type: %s", c.entityLabel, err)
+	}
+
+	vcdClient := meta.(*VCDClient)
+	err = execSchemaHook(vcdClient, d, c.preCreateHooks)
+	if err != nil {
+		return diag.Errorf("error executing pre-create %s hooks: %s", c.entityLabel, err)
+	}
+
+	createdEntity, err := c.createFunc(t)
+	if err != nil {
+		return diag.Errorf("error creating %s: %s", c.entityLabel, err)
+	}
+
+	err = c.stateStoreFunc(d, createdEntity)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return c.resourceReadFunc(ctx, d, meta)
+}
+
+func updateResource[O updateDeleter[O, I], I any](ctx context.Context, d *schema.ResourceData, meta interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	t, err := c.getTypeFunc(d)
+	if err != nil {
+		return diag.Errorf("error getting %s type: %s", c.entityLabel, err)
+	}
+
+	retrievedEntity, err := c.getEntityFunc(d.Id())
+	if err != nil {
+		return diag.Errorf("error getting %s: %s", c.entityLabel, err)
+	}
+
+	_, err = retrievedEntity.Update(t)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return c.resourceReadFunc(ctx, d, meta)
+}
+
+func readResource[O updateDeleter[O, I], I any](_ context.Context, d *schema.ResourceData, _ interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	retrievedEntity, err := c.getEntityFunc(d.Id())
+	if err != nil {
+		if govcd.ContainsNotFound(err) {
+			util.Logger.Printf("[DEBUG] entity '%s' with ID '%s' not found. Removing from state", c.entityLabel, d.Id())
+		}
+		return diag.Errorf("error getting %s: %s", c.entityLabel, err)
+	}
+
+	err = execEntityHook(retrievedEntity, c.readHooks)
+	if err != nil {
+		return diag.Errorf("error executing read %s hooks: %s", c.entityLabel, err)
+	}
+
+	err = c.stateStoreFunc(d, retrievedEntity)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return nil
+}
+
+func readDatasource[O updateDeleter[O, I], I any](_ context.Context, d *schema.ResourceData, _ interface{}, c crudConfig[O, I]) diag.Diagnostics {
```

Comment:
Should we state somewhere that `"name"` argument is required for this generic function to work? (Some kind of pre-requisite for the generic data source approach).

As an alternative, could it be an input parameter?

Reply:
I have put a comment for now. I think what we can do if this is hit as a problem - we could introduce an override field in `crudConfig`- meaning that `name` is probably our default, but if we want - we could modify the expected field name